File: examples/run_batch_model/run_batch_model.cpp
// Copyright (c) 2022 DEEPX Corporation. All rights reserved.
// Licensed under the MIT License.

#include "dxrt/dxrt_api.h"

#include <string>
#include <iostream>


int main(int argc, char* argv[])
{
    const int DEFAULT_LOOP_COUNT = 1;
    const int DEFAULT_BATCH_COUNT = 1;
    
    std::string modelPath;
    int loop_count = DEFAULT_LOOP_COUNT;
    int batch_count = DEFAULT_BATCH_COUNT;

    if ( argc > 2 )
    {
        modelPath = argv[1];

        if ( argc > 2 ) 
        {
            loop_count = std::stoi(argv[2]);
        }

        if ( argc > 3 ) 
        {
            batch_count = std::stoi(argv[3]);
        }

    }
    else
    {
        std::cout << "[Usage] run_batch_model [dxnn-file-path] [loop-count] [batch-count]" << std::endl;
        return -1;
    }

    
    try
    {

        // create inference engine instance with model
        dxrt::InferenceEngine ie(modelPath);

        // create temporary input buffer for example
        std::vector<uint8_t> inputBuffer(ie.GetInputSize(), 0);

        // input buffer vector
        std::vector<void*> inputBuffers;
        for(int i = 0; i < batch_count; ++i)
        {
            // assigns the same buffer pointer in this example
            inputBuffers.emplace_back(inputBuffer.data());
        }
        

        std::cout << std::endl; // blank line
        std::cout << "[output-internal] Use user's output buffers" << std::endl;
        

        // output buffer vector
        std::vector<void*> output_buffers(batch_count, 0);

        // create user output buffers
        for(auto& ptr : output_buffers)
        {
            ptr = new uint8_t[ie.GetOutputSize()];
        } // for i

        std::cout << "[output-user] Create output buffers by user" << std::endl;
        std::cout << "[output-user] These buffers should be deallocated by user" << std::endl;

        auto start = std::chrono::high_resolution_clock::now();

        // batch inference loop
        for(int i = 0; i < loop_count; ++i)
        {
            // inference asynchronously, use all npu core
            auto outputPtrs = ie.Run(inputBuffers, output_buffers);

            std::cout << "[output-user] Inference outputs (" << i << ")" << std::endl;
            std::cout << "[output-user] Inference outputs size=" << outputPtrs.size() << std::endl;
            std::cout << "[output-user] Inference outputs first-tensor-name" << outputPtrs.front().front()->name() << std::endl;

            // postProcessing(outputs);
            (void)outputPtrs;
            std::cout << "[output-user] Reuse the user's output buffers" << std::endl;
        }

        auto end = std::chrono::high_resolution_clock::now();

        // Deallocated the user's output buffers
        for(auto& ptr : output_buffers)
        {
            delete[] static_cast<uint8_t*>(ptr);
        } // for i
        std::cout << "[output-user] Deallocated the user's output buffers" << std::endl;

        std::chrono::duration<double, std::milli> duration = end - start;

        double total_time = duration.count();
        double avg_latency = total_time / static_cast<double>(loop_count*batch_count);
        double fps = 1000.0 / avg_latency;

        std::cout << "---------------------------------------------" << std::endl;
        std::cout << "Use user's output buffers" << std::endl;
        std::cout << "Total Count: loop=" << loop_count << ", batch=" << batch_count << ", total=" << loop_count * batch_count << std::endl;
        std::cout << "Total Time: " << total_time << " ms" << std::endl;
        std::cout << "Average Latency: " << avg_latency << " ms" << std::endl;
        std::cout << "FPS: " << fps << " frames/sec" << std::endl;
        std::cout << "Success" << std::endl;
        std::cout << "---------------------------------------------" << std::endl;
         
    }
    catch (const dxrt::Exception& e)
    {
        std::cerr << e.what() << " error-code=" << e.code() << std::endl;
        return -1;
    }
    catch (const std::exception& e)
    {
        std::cerr << e.what() << std::endl;
        return -1;
    }
    catch(...)
    {
        std::cerr << "Exception" << std::endl;
        return -1;
    }
    
    return 0;
}

File: examples/python/run_batch_model.py
import numpy as np
import sys
from dx_engine import InferenceEngine
from dx_engine import InferenceOption


if __name__ == "__main__":
    DEFAULT_LOOP_COUNT = 1
    DEFAULT_BATCH_COUNT = 1
    
    loop_count = DEFAULT_LOOP_COUNT
    batch_count = DEFAULT_BATCH_COUNT
    modelPath = ""
    argc = len(sys.argv)
    if ( argc > 1 ) :
        modelPath = sys.argv[1];
        if ( argc > 2 ) :
            loop_count = int(sys.argv[2])
        if ( argc > 3 ) :
            batch_count = int(sys.argv[3])
    else:
        print("[Usage] run_batch_model [dxnn-file-path] [loop-count] [batch-count]")
        exit(-1)
   
    print('loop-count=', loop_count)
    print('batch-count=', batch_count)
   
    # create inference engine instance with model
    with InferenceEngine(modelPath) as ie:

        # register call back function
        #ie.register_callback(onInferenceCallbackFunc)

        input_buffers = []
        output_buffers = []
        index = 0
        for b in range(batch_count):
            input_buffers.append([np.zeros(ie.get_input_size(), dtype=np.uint8)])
            output_buffers.append([np.zeros(ie.get_output_size(), dtype=np.uint8)])
            index = index + 1

        # inference loop
        for i in range(loop_count):

            # batch inference
            # It operates asynchronously internally for the specified number of batches and returns the results
            results = ie.run_batch(input_buffers, output_buffers)

            # post processing 
            #postProcessing(outputs)
            print("Inference outputs ", i)
            print("Size of result", len(results))
            for result in results:
                print("Output (Result)", result)

    exit(0)